### PR TITLE
Update CircleCI config changes from CircleCI branch into main

### PR DESCRIPTION
**Pull request by mohavro**
## Description

The purpose of this PR is to update the main branch with the most recent changes to .cricleci/config.yml to the CircleCI branch. 

## Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [X] Code refactoring
- [X] Performance improvement
- [ ] Test addition/update
- [X] Dependency update

## Related Issues

Closes #

## Changes Made

-
-
-

## Testing

- [X] Unit tests pass locally
- [X] Integration tests pass locally
- [ ] Manual testing completed
- [ ] New tests added for new functionality

### Test Commands

```bash
# Commands used to test
pytest
npm test
```

### Code Quality

- [X] My code follows the project's style guidelines
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### Documentation

- [ ] I have updated the README.md (if needed)
- [ ] I have updated the CHANGELOG.md
- [ ] I have updated relevant documentation files
- [X] I have added/updated docstrings

### Dependencies

- [X] I have checked for dependency conflicts
- [X] I have updated requirements.txt (if needed)
- [X] I have updated package.json (if needed)

### Branch Management

- [ ] My branch is up to date with main
- [ ] I will delete this branch after merge
- [X] This branch has a descriptive name
- [ ] This PR has been open for less than 2 weeks

## Additional Notes

**For Reviewers**: Please ensure the branch is deleted after merging this PR to maintain repository hygiene.

**Related Documentation**:
- [Contributing Guidelines](../CONTRIBUTING.md)
- [Branch Cleanup Guidelines](../BRANCH_CLEANUP_QUICK_REFERENCE.md)

---
## Summary by cubic
Updates main with faster CircleCI pipelines (v2 caching with fallbacks, medium executors, native test splitting, parallel jobs) and fully decouples security scans. Restores the schema report with ASCII-only output and tightens CI/config/Snyk tests; also updates dev dependencies.

- **Refactors**
  - CI caching/executors: v2 Python/Node caches with fallback keys; cache site-packages, ~/.npm, and .next/cache; separate python-security cache; set medium resource_class; remove node/python/orb-tools orbs; npm ci uses --prefer-offline and --no-audit.
  - CI workflow: native test splitting; lint/tests and frontend lint/build run in parallel; python-security runs independently of lint; Docker build waits on python-test and python-security; fixed TESTFILES quoting; test glob to tests/**/test_*.py.
  - API/report/tests: login endpoint now sync; schema report uses ASCII-only hyphens and fixed indentation; strengthened config validation and Snyk tests (added raw-content fixture, type hints, unit marks; removed .env.example check).

- **Dependencies**
  - Updated to pyyaml>=6.0.0 and added PyGithub; removed ruamel.yaml, types-PyYAML, Markdown, and typer pins.

<sup>Written for commit e013ed0f8e2b3e60d2cd9ad8c145312c98dc9d43. Summary will update on new commits. <a href="https://cubic.dev/pr/DashFin-FarDb/financial-asset-relationship-db/pull/908">Review in cubic</a></sup>

### **Description**
This PR introduces significant enhancements to the CircleCI configuration and adds unit tests for configuration validation:
- Optimized CircleCI configuration for better performance, including caching improvements and resource allocation.
- Added unit tests to validate CircleCI and other configuration files, ensuring correctness and consistency.
- Refactored existing code for better maintainability and readability.

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config.yml</strong><dd><code>Optimize CircleCI Configuration for Performance</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.circleci/config.yml
<li>Optimized CircleCI configuration for better performance.<br> <li> Added caching improvements for Python and Node dependencies.<br> <li> Implemented test splitting for parallel execution.<br> <li> Enhanced resource allocation for executors.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/908/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47">+53/-33</a>&nbsp; </td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>suggest_fixes.py</strong><dd><code>Refactor Code Suggestion Extraction Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/pr-copilot/scripts/suggest_fixes.py
<li>Refactored code for extracting code suggestions.<br> <li> Improved readability and efficiency of the function.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/908/files#diff-f0de8f989e9f3a5cf5504706b540be07f04f1fd5bb3f47adad5d38a5de763dce">+4/-12</a>&nbsp; &nbsp; </td>
</tr>
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_config_validation.py</strong><dd><code>Add Unit Tests for Configuration Validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_config_validation.py
<li>Added unit tests for validating CircleCI configuration.<br> <li> Improved existing tests for configuration files.<br> <li> Enhanced test coverage for various configurations.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/908/files#diff-59de2007b696f693118300a56df156429a432f41c2fc7169c79ce898ad401f50">+1293/-818</a></td>
</tr>
</table></td></tr></tr></tbody></table>

> 💡 **Penify usage**:
>Comment `/help` on the PR to get a list of all available Penify tools and their descriptions

___

## **CodeAnt-AI Description**
Add comprehensive configuration and CI validation tests, fix report generation, and update dev dependencies

### What Changed
- Rewrote and expanded unit tests for repository configuration files: added CircleCI, GitHub Actions, Copilot, Codacy and CircleCI edge-case validations and strengthened existing checks (vercel.json, package.json, tsconfig, .env.example, .gitignore, requirements.txt, PostCSS, Tailwind).
- Added YAML parsing fixtures and new tests that validate .circleci/config.yml structure, executor/resource classes, cache/restore consistency, workflow job dependencies, and scheduling for nightly security scans.
- Added tests and content fixtures for GitHub composite actions, Copilot instructions, issue templates, Codacy docs, and Snyk workflow raw-content checks to detect broken links, missing fields, and hardcoded values.
- Fixed schema report generator so it correctly collects metrics and returns a well-formed Markdown report (restored and cleaned rendering helpers and header/section formatting).
- Cleaned up small formatting/whitespace issues across code (API endpoint parameter formatting preserved behavior) and updated requirements-dev.txt to replace some pins with current packages (pyyaml, PyGithub) and remove obsolete entries.

### Impact
`✅ Clearer CI config validation preventing broken pipelines`
`✅ Fewer CI regressions by catching workflow/orb/executor issues early`
`✅ Clearer, consistently generated schema reports`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/81d4f49d-7a10-4144-951a-6739490c8d7b?refs=0154db3af30a2d197b907a6f6bb53bb9ba93f4ad&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: .circleci/config.yml [hunk 1]
```diff
@@ -59,21 +68,15 @@ jobs:
...
-            pytest \
-              --cov=src \
-              --cov=tests \
-              --cov-report=xml \
-              --cov-report=html \
-              --cov-report=term-missing \
-              -v \
-              --junitxml=test-results/pytest/results.xml
+            TESTFILES=$(circleci tests glob "tests/**/*.py" | circleci tests split --split-by=timings)
+            pytest "$TESTFILES" --cov=src --cov=tests --cov-report=xml --cov-report=html --cov-report=term-missing -v --junitxml=test-results/pytest/results.xml
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .circleci/config.yml]: This is intentional to run parellism?

File: tests/unit/test_config_validation.py [hunk 2]
```diff
@@ -1,818 +1,1368 @@
...
+        # Extract frontmatter
+        lines = custom_issue_template.split("\n")
+        if lines[0] == "---":
+            # Find end of frontmatter
+            end_idx = lines[1:].index("---") + 1
+            frontmatter = "\n".join(lines[1:end_idx])
+
+            # Should have name and about fields
+            assert "name:" in frontmatter
+            assert "about:" in frontmatter
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_config_validation.py]:
```suggestion
    def test_custom_issue_template_frontmatter_fields(self, custom_issue_template):
        """Test that frontmatter has required fields."""
        # Extract frontmatter
        lines = custom_issue_template.split("\n")
        if lines[0] == "---":
            # Find end of frontmatter
            try:
                end_idx = lines[1:].index("---") + 1
                frontmatter = "\n".join(lines[1:end_idx])

                # Should have name and about fields
                assert "name:" in frontmatter
                assert "about:" in frontmatter
            except ValueError:
                pytest.fail("Frontmatter closing delimiter '---' not found")
```

File: tests/unit/test_config_validation.py [hunk 3]
```diff
@@ -1,818 +1,1369 @@
...
+        config_path = Path("frontend/next.config.js")
+        assert config_path.exists(), "next.config.js not found"
+
+        with open(config_path) as f:
+            return f.read()
+
+    def test_next_config_exists(self) -> dict:
+        """Test that next.config.js exists."""
+        config_path = Path("frontend/next.config.js")
+        assert config_path.exists()
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_config_validation.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/908#discussion_r2861795193)

**Review by mohavro — COMMENTED** [review]
approved
